Fix logistic regression model

TomasVaskevicius opened the pull request.

TomasVaskevicius commented on firefly/DESCRIPTION:
```diff
@@ -2,8 +2,8 @@ Package: firefly
 Authors@R: c(
-  person("Francesca", "Crucinio"),
-  person("Tomas", "Vaskevicius"))
+  person("Francesca", "Crucinio", email = "francesca.crucinio@spc.ox.ac.uk", role = c("aut")),
```
I had to add our emails, otherwise devtools wouldn't work.

TomasVaskevicius commented on firefly/R/logistic_regression.R:
```diff
@@ -1,99 +1,107 @@
-    xi <- t(map.tuned)%*%(x*t)
-    }
-   
-    a <- -(exp(xi)-1)/(4*xi*(exp(xi)+1))
```
I think there was a mistake in the paper when computing the a. 
Check out the https://pdfs.semanticscholar.org/e407/ea7fda6d152d2186f4b5e27aa04ec2d32dcd.pdf.
My change is consistent with how they compute "a" in their python code I think.

TomasVaskevicius commented on firefly/R/logistic_regression.R:
```diff
@@ -1,99 +1,107 @@
+  if (theta.map.estimate != NULL) {
+    # We can come up with a better value of xi.
+    # Implement this later.
+    stop(paste("Using theta.map.estimate int logistic regression model",
```
Let's only worry about tuned parameters once we have something working.

FrancescaCrucinio commented on firefly/R/logistic_regression.R:
```diff
@@ -1,99 +1,107 @@
+    #        evaluated.
+    # @return A vector of log likelihoods for the requested subset of our
+    #         dataset.
+    log(1 / (1 + exp((X[ids,] %*% theta) * y[ids])))
```
There's a - missing 
built in function inv.logit would work faster

TomasVaskevicius replied:
You are right! Will fix in a minute.

TomasVaskevicius commented on firefly/R/logistic_regression.R:
```diff
@@ -1,99 +1,107 @@
-  ##Compute log pseudo prior
-  LogPseudoPrior <- function(x, theta, map.tuned, scale, t){ 
-    log.prior <- sum(dnorm(theta, mu, scale, log=TRUE)) #compute log prior
-    log.bound <- LogBound(x, theta, map.tuned, t)
```
This is too slow, there is a more effiient way to compute the log product of bounds for all the dataset! This is precisely why the bound functions have to be chosen in a way to make the computation of their product over all of the dataset quick.

TomasVaskevicius commented on firefly/R/logistic_regression.R:
```diff
@@ -1,99 +1,107 @@
+    log(likelihoods - exp(LogBound(ids, theta)))
   }
-  
-  LogGradient <-function(x, theta, map.tuned, z, t){
```
I have removed this for a while. Looks way too complicated and we first need to focus on getting something to work. After that we will put this function back and test it.

FrancescaCrucinio commented on firefly/R/logistic_regression.R:
```diff
@@ -1,99 +1,107 @@
+  # Define a defaul vector xi, specifying where each log likelihood lower bound
+  # should be tight.
+  xi <- rep(1.5, nrow(X))
+  if (theta.map.estimate != NULL) {
```
Robin Evans suggested to use is.null() instead of !=

FrancescaCrucinio commented on firefly/R/logistic_regression.R:
```diff
@@ -1,99 +1,107 @@
+  a <- (sig.neg.xi - sig.xi) / (4 * xi)
+  b <- 1/2
+  c <- -a * xi^2 - xi / 2 + log(sig.xi)
+
```
There are 2 version for computing a =O
paper and phd thesis give the same value (the one i used), but Python code gives another one
no idea of which is correct

FrancescaCrucinio commented on firefly/R/logistic_regression.R:
```diff
@@ -1,99 +1,107 @@
+    #        evaluated.
+    # @return A vector of log differences for the requested subset of our
+    #         dataset.
+    likelihoods <- 1 / (1 + exp((X[ids,] %*% theta) * y[ids]))
```
see above (- missing)

FrancescaCrucinio commented on firefly/R/logistic_regression.R:
```diff
@@ -1,99 +1,107 @@
+    # @return A vector of log bounds for the requested subset of our
+    #         dataset.
+    X.theta.y <- (X[ids,] %*% theta) * y[ids]
+    a[ids] * (X.theta.y)^2 + b[ids] * X.theta.y + c[ids]
```
I don't think the indexes in b are needed, +b should work fine

FrancescaCrucinio commented on firefly/R/logistic_regression.R:
```diff
@@ -1,99 +1,107 @@
+    #        evaluated.
+    # @return A vector of log likelihoods for the requested subset of our
+    #         dataset.
+    log(plogis((X[ids,] %*% theta) * y[ids]))
```
is this invers logit? I think is just logit

TomasVaskevicius commented:
Hi Francesca,

I have fixed the likelihood computation to use the native code (inv.logit function that you suggested just calls plogis, so this is what I do now).

Also changed the != NULL like to !is.null()
